floris/tools/layout_functions.py

```diff
@@ -397,18 +344,232 @@ def put_label(i, ax):
+        # D = np.array([turb['rotor_diameter'] for turb in
+        #      fi.floris.farm.turbine_definitions])
+    # else:
+    # D = D*np.ones(N_turbs)
+
+    dists_m = np.zeros((N_turbs, N_turbs))
+    angles_d = np.zeros((N_turbs, N_turbs))
+
+    for i in range(N_turbs):
+        for j in range(N_turbs):
+            dists_m[i, j] = np.linalg.norm([layout_x[i] - layout_x[j], layout_y[i] - layout_y[j]])
+            angles_d[i, j] = get_wake_direction(layout_x[i], layout_y[i], layout_x[j], layout_y[j])
+
+    # Mask based on the limit distance (assumed to be in measurement D)
+    if limit_dist_D is not None and limit_dist_m is None:
+        limit_dist_m = limit_dist_D * D
+    if limit_dist_m is not None:
+        mask = dists_m > limit_dist_m
+        dists_m[mask] = np.nan
+        angles_d[mask] = np.nan
+
+    # Handle default limit number case
+    if limit_num is None:
+        limit_num = -1
+
+    # Loop over pairs, plot
+    label_exists = np.full((N_turbs, N_turbs), False)
+    for i in range(N_turbs):
+        for j in range(N_turbs):
+            # import ipdb; ipdb.set_trace()
+            if (
+                ~np.isnan(dists_m[i, j])
+                and dists_m[i, j] != 0.0
+                and ~(dists_m[i, j] > np.sort(dists_m[i, :])[limit_num])
+                # and i in layout_plotting_dict["turbine_indices"]
+                # and j in layout_plotting_dict["turbine_indices"]
+            ):
+                (h,) = ax.plot(fi.layout_x[[i, j]], fi.layout_y[[i, j]], **wake_plotting_dict)
+
+                # Only label in one direction
+                if ~label_exists[i, j]:
+                    linetext = "{0:.1f} D --- {1:.0f}/{2:.0f}".format(
+                        dists_m[i, j] / D,
+                        angles_d[i, j],
+                        angles_d[j, i],
+                    )
+
+                    label_line(
+                        h,
+                        linetext,
+                        ax,
+                        near_i=1,
+                        near_x=None,
+                        near_y=None,
+                        rotation_offset=0,
+                        size=wake_label_size,
+                    )
+
+                    label_exists[i, j] = True
+                    label_exists[j, i] = True
+
+    return ax
+
+
+def plot_farm_terrain(fi: FlorisInterface, fig: plt.Figure, ax: plt.Axes) -> None:
+    """
+    Creates a filled contour plot visualizing terrain-corrected wind turbine hub heights.
+
+    Args:
+        fi (FlorisInterface): Instantiated FlorisInterface object containing layout data.
+        fig (plt.Figure):  The figure object where the plot will be created.
+        ax (plt.Axes): The axes object where the plot will be drawn.
+    """
+    hub_heights = fi.floris.farm.hub_heights.flatten()
+    cntr = ax.tricontourf(fi.layout_x, fi.layout_y, hub_heights, levels=14, cmap="RdBu_r")
+
+    fig.colorbar(
```

Comment:
This is the only function in this module that operates on the figure. Consider refactoring for consistency with the others, if possible.

Reply:
It's possible, but colorbar is a finnicky one, I think that is why it's attached to the figure, an apparant alternative would be:

```
import matplotlib.pyplot as plt
from matplotlib.colorbar import Colorbar

# Assuming cntr is a contour plot
cbar = Colorbar(ax=ax, mappable=cntr, orientation='vertical', ticks=np.linspace(
            np.min(hub_heights) - 10.0,
            np.max(hub_heights) + 10.0,
            15,
        ))
cbar.set_label("Terrain-corrected hub height (m)")
```

Reply:
I've updated this so that `fig` is just grabbed from the passed `ax` using `ax.get_figure()`. That removes the need for `fig` to be passed separately. I've also added to the 23 example to check that works and see no difference in output between the original implementation (which received `fig`) and the change (which no longer requires/accepts `fig` as an argument).

Reply:
Thank you @misi9170 !